clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2672,6 +2949,87 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
     addFieldInit(FD, Ty, ParamRef);
   }
 
+  Expr *addDerivedToBaseCastExpr(const CXXRecordDecl *RD,
+                                 const CXXBaseSpecifier &BS,
+                                 Expr *LocalCloneRef) {
+    CXXCastPath BasePath;
+    QualType DerivedTy(RD->getTypeForDecl(), 0);
+    QualType BaseTy = BS.getType();
+    SemaRef.CheckDerivedToBaseConversion(DerivedTy, BaseTy, KernelCallerSrcLoc,
+                                         SourceRange(), &BasePath,
+                                         /*IgnoreBaseAccess*/ true);
+    auto Cast = ImplicitCastExpr::Create(
+        SemaRef.Context, SemaRef.Context.getPointerType(BaseTy),
+        CK_DerivedToBase, LocalCloneRef,
+        /* CXXCastPath=*/&BasePath, VK_LValue, FPOptionsOverride());
+    return Cast;
+  }
+
+  Expr *getAddressOf(Expr *E) {
+    return UnaryOperator::Create(SemaRef.Context, E, UO_AddrOf,
+                                 SemaRef.Context.getPointerType(E->getType()),
+                                 VK_PRValue, OK_Ordinary, KernelCallerSrcLoc,
+                                 false, SemaRef.CurFPFeatureOverrides());
+  }
+
+  Expr *buildMemCpyCall(Expr *From, Expr *To, QualType T) {
+    // Compute the size of the memory buffer to be copied.
+    QualType SizeType = SemaRef.Context.getSizeType();
+    llvm::APInt Size(SemaRef.Context.getTypeSize(SizeType),
+                     SemaRef.Context.getTypeSizeInChars(T).getQuantity());
+
+    LookupResult R(SemaRef, &SemaRef.Context.Idents.get("__builtin_memcpy"),
+                   KernelCallerSrcLoc, Sema::LookupOrdinaryName);
+    SemaRef.LookupName(R, SemaRef.TUScope, true);
+
+    FunctionDecl *MemCpy = R.getAsSingle<FunctionDecl>();
+
+    assert(MemCpy && "__builtin_memcpy should be found");
+
+    ExprResult MemCpyRef =
+        SemaRef.BuildDeclRefExpr(MemCpy, SemaRef.Context.BuiltinFnTy,
+                                 VK_PRValue, KernelCallerSrcLoc, nullptr);
+
+    assert(MemCpyRef.isUsable() && "Builtin reference cannot fail");
+
+    Expr *CallArgs[] = {To, From,
+                        IntegerLiteral::Create(SemaRef.Context, Size, SizeType,
+                                               KernelCallerSrcLoc)};
+    ExprResult Call =
+        SemaRef.BuildCallExpr(/*Scope=*/nullptr, MemCpyRef.get(),
+                              KernelCallerSrcLoc, CallArgs, KernelCallerSrcLoc);
+
+    assert(!Call.isInvalid() && "Call to __builtin_memcpy cannot fail!");
+    return Call.getAs<Expr>();
+  }
+
+  // Adds default initializer for generated type and creates
+  // a call to __builtin_memcpy to iniatilize local clone from
+  // kernel argument.
+  void handleGeneratedType(FieldDecl *FD, QualType Ty) {
+    addFieldInit(FD, Ty, None,
```

Comment:
This is not working when the field type is a lambda class. For example -

```
float *Ptr = &A;
  auto Lambda = [=]() {
    *Ptr += 1;
  };
  cl::sycl::kernel_single_task<class lambdas>([=]() {
    Lambda();
  });

```
Here the kernel object field is a lambda class. The visitor visits this record like any other class, sees it has a pointer and marks it for 'new type' generation. The issue is lambda classes do not have default constructors when the lambda has captures (https://clang.llvm.org/doxygen/DeclCXX_8cpp_source.html#l00676). So I'm not sure how to handle this.  I created a default initialization here so the integrity of the InitListExpr is maintained (like we do for accessors). Actual field initialization takes place in KernelBody via memcpy.

@erichkeane @Fznamznon @premanandrao any ideas on how to handle this?

Reply:
This worked with the old logic because it used copy constructor (to initialize from kernel parameter). Since the kernel parameter is now a generated new type, we can't do this and hence the default init + memcopy

Reply:
I skipped decomposition of lambdas for now. I'm not sure if this is a case we need to handle and we can revisit it if we need to do so.